### PR TITLE
Update pre-processor logic for VJ Amp includes

### LINKED ISSUE #5989: If VJ Include has an AMP alternative, ensure we fetch it and make it available
**Is your feature request related to a problem? Please describe.**
So far we do not render any VJ content on AMP pages. However more recent (c. 2018-) VJ includes have an AMP version available (the so-called 'AMP wrapper') on the `/amp` endpoint.

**Describe the solution you'd like**
Where a VJ include is known to have an AMP version, we should request it and add it to the Include block so that it's available to the page. Otherwise the block should be omitted.

**Possible routes**
1. See additional context below on `amp-image` query parameter.
2. Can Ares decorate existing Include data to tell us whether there is an AMP version available?
3. If we cannot determine the Include has an AMP version (i.e. confirm it was published after 2018) _before_ fetching it - we may have to fetch it from the `/amp` and determine its availability through HTTP status codes. 404/200.

Note: We should keep unnecessary network requests to a minimum.

**Describe alternatives you've considered**

**Testing notes**
- Need to test this with a mixture of VJ content old and new to ensure it is working as expected across canonical and AMP.

[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [x] This feature is expected to need manual testing.

**Additional context**
- According to @eagerterrier the presence of the `amp-image` query parameter on the include path is expected to be a reliable way of determining if an AMP version is available.
- The pathname may be available to us, if we want to check which platform (AMP/Canonical) we are in, inside the pre-processor: https://github.com/bbc/simorgh/pull/6758.

### PR DESCRIPTION
**Pull request by andrewscfc**
Resolves https://github.com/bbc/simorgh/issues/5989

**Overall change:**
Changes our pre-processor logic for includes to add `ampSrc` attribute to our include block for use in the frontend as the `src` for our `amp-iframe`. The `html` attribute is omitted as we do not need to fetch the html markup as the `amp-iframe` will perform this fetch in the users browser.

```javascript
{
      type: 'include',
      model: {
        href:  '/include/newsspec/21841-green-diet/gahuza/app?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png',
        required: false,
        tile: 'Include from VisJo',
        platform: 'highweb',
        type: 'vj',
        ampSrc:
          'https://news.files.bbci.co.uk/include/newsspec/21841-green-diet/gahuza/app/amp?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png',
}
```

Additionally, we omit the include block entirely for VJ includes on AMP that do not have AMP support, this maintaining our existing behaviour to show no include on AMP where there is no support; a fallback experience follows in a subsequent issue.

**Code changes:**
- This change includes an opportunistic refactor to introduce a `includeClassifier` component, this gives includes a `classification` this used to determine how the block is pre-processed in the include converter. It extracts the logic to determine what type the include is and encapsulates new logic that determines whether VJ includes support AMP or not.
- `ampSrcBuilder` is introduced to build the `src` attribute for an amp iframe, adding `/amp` to the path before any get parameters
- The `includeClassifier` is integrated into the include converter, and used to determine whether to:
  - add the `ampSrc` attribute to the block data, `ampSrcBuilder` is used to build the url.
  - add the `html` attirbute to the block data and request the include payload
  - to omit the block entirely where the VJ include is not supported on AMP

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [ ] I have assigned this PR to the Simorgh project

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js [hunk 1]
```diff
@@ -88,34 +81,38 @@ const convertInclude = async (includeBlock, ...restParams) => {
...
-  // This extracts the type from the href
-  const typeExtraction = Object.keys(supportedTypes).find(
-    hrefIsSupported(href),
-  );
+  const { includeType, classification } = includeClassifier({ href, pathname });
 
-  // This determines if the type is supported and returns the include type name
-  const includeType = supportedTypes[typeExtraction];
-  if (!includeType) {
+  if (
```
Comment (rhenshaw56) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: this reads fine but at the time I saw this initially, there was 2 if blocks, not sure if this reads better
```
  const unsupportedLogClassifier = {
    'not-supported': INCLUDE_UNSUPPORTED,
    'vj-include-not-supporting-amp': INCLUDE_UNSUPPORTED,
  };

  if (unsupportedLogClassifier[classification]) {
    logger.info(unsupportedLogClassifier[classification], {
      type,
      classification,
      url: href,
    });
    return null;
  }
```

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js [hunk 2]
```diff
@@ -88,34 +81,38 @@ const convertInclude = async (includeBlock, ...restParams) => {
...
+  let html;
+  if (classification === 'vj-include-supports-amp') {
+    ampSrc = ampSrcBuilder(href);
+    logger.info(INCLUDE_REQUEST_RECEIVED, {
+      url: ampSrc,
+    });
+  } else {
+    html = await fetchMarkup(buildIncludeUrl(href, includeType, pathname));
+  }
+
```
Comment (rhenshaw56) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:113]: wonder if we could return this implicitly with a tenary
```
  return {
    type,
    model: {
      href,
      type: includeType,
      ...(classification === 'vj-include-supports-amp'
        ? {
            ampSrc: ampSrcBuilder(href),
          }
        : {
            html: await fetchMarkup(
              buildIncludeUrl(href, includeType, pathname),
            ),
          }),
      ...rest,
    },
  };
```

Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:113]: this way we won't have `html` or `ampSrc` being undefined in the include block, they won't be there at all

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js [hunk 3]
```diff
@@ -94,7 +102,7 @@ describe('convertInclude', () => {
     });
   });
 
-  it('should fetch and convert an include block to a vj block', async () => {
+  it('should fetch an convert an include block to a vj block', async () => {
```
Comment (rhenshaw56) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]:
```suggestion
  it('should fetch an convert and include block to a vj block', async () => {
```

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js (hunk 2 above)
Comment (Simon2828) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:113]: Should I move the logging into `ampSrcBuilder`?

Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:113]: I think because we `ampSrc` and `html` are `undefined` in the relevant code paths, they are not added to the object, we check for that in our tests:
https://github.com/bbc/simorgh/pull/6873/files#diff-754f58c9944064fefc2f77cd295825f9R325-R336 

I think personally think it is more readable in its current form, but interested to see what others think?

Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js:113]: > Should I move the logging into ampSrcBuilder? 

I'd suggest leaving logging in the converter as our focused libraries probably shouldn't take on the extra concern of logging, better to keep them [pure](https://medium.com/better-programming/what-is-a-pure-function-3b4af9352f6f) so they don't have any side-effects such as logging.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+    expect(actual).toEqual(expected);
+  });
+
+  it.each`
+    includeType | classification            | href
+    ${`idt1`}   | ${`idt1-canonical`}       | ${`indepthtoolkit/quizzes/123-456`}
+    ${`idt2`}   | ${`idt2-canonical`}       | ${`idt2/111-222-333-444-555`}
+    ${`vj`}     | ${`vj-include-canonical`} | ${`include/newsspec/21841-green-diet/gahuza/app?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png`}
+  `(
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js:119]: I've used a for loop in the past, this looks a much a neater approach 👍

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  '/include/newsspec/21841-green-diet/gahuza/app?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png';
+const vjIncludeNotSupportingAmp =
+  '/news/special/2016/newsspec_14813/content/iframe/gahuza/us-gop.inc?responsive=true&app-clickable=true&app-image=http://a.files.bbci.co.uk/worldservice/live/assets/images/2016/11/09/161109092836_us_election_2nddaymaps_winner_ws_62_v3.png';
+const idt1Include = '/indepthtoolkit/quizzes/123-456';
+const idt2Include = '/idt2/111-222-333-444-555';
+
+const canonicalPathname = 'https://www.bbc.com/service/foo';
+const ampPathname = 'https://www.bbc.com/service/foo.amp';
+
+describe('Include Classifier', () => {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js:13]: Maybe we can group some the tests together in this file using the `it` approach you've found below? So maybe:
- A test for all include types on AMP
- A test for all include types on canonical

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js (hunk 4 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js:119]: May want to add tests for the leading slash for amp too? Be careful not to overcomplicate this test though, better to duplicate test for Amp than have more and more rows in your table above.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js (hunk 1 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: Another variant 
```javascript
const unsupportedClassifications = [ 'not-supported', 'vj-include-not-supporting-amp' ];
if(unsupportedClassifications.includes(classification)) {
...
```

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    return { includeType, classification: 'vj-include-not-supporting-amp' };
+  }
+  if (includeType) {
+    const platform = pathnameIsAmp ? 'amp' : 'canonical';
+    const classification = `${includeType}-${platform}`;
+
+    return { includeType, classification };
+  }
+  return { includeType: null, classification: 'not-supported' };
```
Comment (rhenshaw56) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js]: Not sure if I missed something but how is the include block being returned differently for `not-supported` and `vj-include-not-supporting-amp`, I reckon if this could be simplified as such:
 
```
  const platform = pathnameIsAmp ? 'amp' : 'canonical';
  const classification = `${includeType}-${platform}`;

  if (!pathnameIsAmp || !ampSupported(href)) {
    return { includeType, classification };
  }

  if (ampSupported(href)) {
    return { includeType, classification: 'vj-include-supports-amp' };
  }

  return { includeType: includeType || null, classification: 'not-supported' };
```

we could group `not-supported` and `vj-include-not-supporting-amp` both as `not-supported`

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js [hunk 7]
```diff
@@ -88,34 +81,38 @@ const convertInclude = async (includeBlock, ...restParams) => {
...
   }
 
+  let ampSrc;
+  let html;
+  if (classification === 'vj-include-supports-amp') {
+    ampSrc = ampSrcBuilder(href);
+    logger.info(INCLUDE_REQUEST_RECEIVED, {
+      url: ampSrc,
+    });
+  } else {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: Maybe move the line from the fetch code above (https://github.com/bbc/simorgh/pull/6873/files#diff-ace5f927a44cc0bc96a151479cba4bbaR59) to this else block:
```javascript
const url = buildIncludeUrl(href, includeType, pathname);
logger.info(INCLUDE_REQUEST_RECEIVED, {
   url,
});
html = await fetchMarkup(buildIncludeUrl(href, includeType, pathname));
```
Just keeps logs of a similar intent close by which _I think_ makes it more readable?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js [hunk 8]
```diff
@@ -88,34 +81,38 @@ const convertInclude = async (includeBlock, ...restParams) => {
...
       url: href,
     });
     return null;
   }
 
+  let ampSrc;
+  let html;
+  if (classification === 'vj-include-supports-amp') {
+    ampSrc = ampSrcBuilder(href);
+    logger.info(INCLUDE_REQUEST_RECEIVED, {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]:
```suggestion
    logger.info(INCLUDE_IFRAME_REQUEST_RECEIVED, {
```
Nice addition of the log, you may want to differentiate this log from `INCLUDE_REQUEST_RECEIVED`, that log implies a request is going to be made to the includes endpoint on server, this log says you are building a url to be used in an iframe on the client.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js (hunk 6 above)
Comment (Simon2828) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js]: For the `vj-include-not-supporting-amp` classification the `includeType` will be `vj` - so it should work on canonical just not on amp. Whereas the `not-supported` classification is for includes which are not included in `supportedTypes` and so won't work on amp or canonical, so I think it is worth keeping the distinction.

_Andrew explained the reasoning so now understand it makes sense to follow your suggestion_ @rhenshaw56  👍

Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js]: That sounds a better approach @rhenshaw56, we don't treat `vj-include-not-supporting-amp` any different to `not-supported` at the converter level. Also will simplify the 'not supported logic' we've discussed here: https://github.com/bbc/simorgh/pull/6873#discussion_r440227803

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/ampSrcBuilder.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import Url from 'url-parse';
+
+const ampSrcBuilder = href => {
+  const ampSrcPrefix = 'https://news.files.bbci.co.uk';
```
Comment (HarveyPeachey) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/ampSrcBuilder.js]: Do we need to add a variant for `https://news.test.files.bbci.co.uk`?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js [hunk 10]
```diff
@@ -88,34 +81,38 @@ const convertInclude = async (includeBlock, ...restParams) => {
...
-    hrefIsSupported(href),
-  );
+  const { includeType, classification } = includeClassifier({ href, pathname });
 
-  // This determines if the type is supported and returns the include type name
-  const includeType = supportedTypes[typeExtraction];
-  if (!includeType) {
+  if (
+    classification === 'not-supported' ||
+    classification === 'vj-include-not-supporting-amp'
```
Comment (HarveyPeachey) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: Just for my understanding. If we have a VJ include that doesn't support amp then we return `null` for canonical and amp? Or am I reading it wrong?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js (hunk 4 above)
Comment (HarveyPeachey) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js:119]: This method looks great!

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/ampSrcBuilder.js (hunk 9 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/ampSrcBuilder.js]: Yeah, we should definitely support the test variant, we should add this to our .env and vary by environment: https://github.com/bbc/simorgh/blob/latest/envConfig/test.env#L7

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js (hunk 10 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: In the case of a VJ include on canonical, it would be classified as `vj-include-canonical` and its html would be fetched below: https://github.com/bbc/simorgh/pull/6873/files/d749bfe55b509c663511be75c1252f5412436506#diff-ace5f927a44cc0bc96a151479cba4bbaR106 

I think @Simon2828 is rolling` 'vj-include-not-supporting-amp'` into `'not-supported'` to make things a bit clearer

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    ? `${includeType}-${platform}`
+    : 'not-supported';
+
+  if (!pathnameIsAmp || (!ampSupported(href) && classification !== 'vj-amp')) {
+    return { includeType, classification };
+  }
+
+  if (ampSupported(href)) {
+    return { includeType, classification: 'vj-include-supports-amp' };
+  }
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js:47]: I find this a bit tricky to read as we have a lot of negative conditionals above, maybe move this if statement up so you can test for:
`if(includeType === 'vj' && ampSupported(href))`

Then you can positively test for each condition and return the correct classfication object. In my head I when I read `if(includeType === 'vj' && ampSupported(href))` I hear "when include type is VJ and it supports amp, return this" as opposed to "when the url is not an amp page, the include does not support amp and it has the classification 'vj-amp', return this" for `!pathnameIsAmp || (!ampSupported(href) && classification !== 'vj-amp')`.

Sorry for the waffle! Just wanted to get across my reasoning for avoiding negations and that's the best I've got 😉

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    });
+
+    expect(actual).toEqual(expected);
+  });
+
+  it.each`
+    includeType | classification               | href
+    ${`idt1`}   | ${`idt1-amp`}                | ${idt1Include}
+    ${`idt2`}   | ${`idt2-amp`}                | ${idt2Include}
+    ${`vj`}     | ${`vj-include-supports-amp`} | ${vjIncludeSupportingAmp}
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js]:
```suggestion
    ${`vj`}     | ${`vj-supports-amp`} | ${vjIncludeSupportingAmp}
```
I did this originally but maybe drop the word `include` here to be consistent with the other classifications

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js (hunk 5 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.test.js:13]: These tests look great to me, thanks for taking the time, a love the `it.each` feature!

**Review by andrewscfc — COMMENTED** [review]
Great changes @Simon2828

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/ampSrcBuilder.test.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import ampSrcBuilder from './ampSrcBuilder';
+
+const includeSupportingAmp =
+  '/include/newsspec/21841-green-diet/gahuza/app?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png';
+
+describe('Amp Src Builder', () => {
+  process.env.SIMORGH_INCLUDES_BASE_AMP_URL = 'https://news.files.bbci.co.uk';
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/ampSrcBuilder.test.js]: This should be cleaned up at the end of the test as it is a change to the global scope that could have side-effects for other tests in the test suite.

Does this have an existing value before the test runs? If not, best to delete it: `delete process.env.   process.env.SIMORGH_INCLUDES_BASE_AMP_URL;` after the test completes.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js (hunk 11 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js:47]: Thanks for addressing the reads really well now 💯

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js [hunk 14]
```diff
@@ -369,7 +252,6 @@ describe('convertInclude', () => {
       model: {
         href: 'idt2',
         type: 'idt2',
-        html: null,
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:372]: Sorry I missed this @rhenshaw56, thanks for suggesting this change: https://github.com/bbc/simorgh/pull/6873/files#diff-ace5f927a44cc0bc96a151479cba4bbaR119-R120

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js [hunk 15]
```diff
@@ -381,15 +263,17 @@ describe('convertInclude', () => {
...
+    );
     expect(fetch).toHaveBeenCalledTimes(1);
     expect(fetch).toHaveBeenCalledWith(
       'https://foobar.com/includes/idt2/html',
       {
         timeout: 3000,
       },
     );
-    expect(loggerMock.info).not.toHaveBeenCalled();
+    expect(loggerMock.info).toHaveBeenCalledTimes(1);
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:282]: Should we assert what info is called with here?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js [hunk 16]
```diff
@@ -533,14 +401,72 @@ describe('convertInclude', () => {
...
+      platform: 'highweb',
+      type: 'include',
+    };
+    const expected = {
+      type: 'include',
+      model: {
+        href: includeSupportingAmp,
+        type: 'vj',
+        ampSrc:
+          'https://foobar.com/includes/amp/include/newsspec/21841-green-diet/gahuza/app/amp?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png',
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: `https://foobar.com/includes/amp/include`, we shouldn't have `/amp` inserted between `/includes` and `/include`, I think we have a bug somewhere, probably in `ampSrcBuilder`. Maybe make a similar test case to this for `ampSrcBuilder` and see what you can find?

**Review by andrewscfc — COMMENTED** [review]
Looking pretty good, just found a bug though identified in one of your tests: https://github.com/bbc/simorgh/pull/6873#discussion_r441540989

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js (hunk 16 above)
Comment (Simon2828) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: I think I just made a mistake in this test file setting the `process.env.SIMORGH_INCLUDES_BASE_AMP_URL` incorrectly rather than it being a bug in `ampSrcBuilder`?
https://github.com/bbc/simorgh/pull/6873/files?file-filters%5B%5D=.js#diff-754f58c9944064fefc2f77cd295825f9R26

Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: I don't think it's a bug. The href begins with a `/include` and the mock for process.env.SIMORGH_INCLUDES_BASE_AMP_URL is `https://foobar.com/includes/amp/include` which combines to result in the weird format, we could maybe remove the `/include` in `includeSupportingAmp`

```
const includeSupportingAmp =
  '/include/newsspec/...
```

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js [hunk 17]
```diff
@@ -15,9 +16,14 @@ const idt2Markup = `<div>IDT 2 Markup</div><script type="text/javascript" src="l
...
 const idt1Markup = `<div>IDT 1 Markup</div><script type="text/javascript" src="localhost/idt1.js"></script>`;
 
+const canonicalPathname = 'https://www.bbc.com/service/foo';
+
+const ampPathname = 'https://www.bbc.com/service/foo.amp';
+
 describe('convertInclude', () => {
   const includesBaseUrl = 'https://foobar.com/includes';
   process.env.SIMORGH_INCLUDES_BASE_URL = includesBaseUrl;
+  process.env.SIMORGH_INCLUDES_BASE_AMP_URL = `${includesBaseUrl}/amp`;
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]:
```suggestion
  process.env.SIMORGH_INCLUDES_BASE_AMP_URL = `https://foobar-amp.com/includes`;
```
To fix the stange path issue

Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: Also these `process.env` vars may want clearing up as mentioned here: https://github.com/bbc/simorgh/pull/6873/files#r441534318

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js (hunk 16 above)
Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: Yeah doesn't look like a bug, maybe just change the base url to differentiate amp from non-amp like this: https://github.com/bbc/simorgh/pull/6873/files#r442018884 

In any case looks like it's just the test 😄

**Review by andrewscfc — COMMENTED** [review]
LGTM, I can't (and shouldn't) approve this it's my original PR and I contributed a lot to it. Great work @Simon2828

**Review by rhenshaw56 — APPROVED** [review]
Alt Text

File: envConfig/live.env [hunk 18]
```diff
@@ -5,6 +5,7 @@ SIMORGH_PUBLIC_STATIC_ASSETS_PATH=/include/articles/public/
 SIMORGH_ATI_BASE_URL=https://a1.api.bbc.co.uk/hit.xiti?
 SIMORGH_TOGGLES_URL=https://toggles.api.bbci.co.uk
 SIMORGH_INCLUDES_BASE_URL=https://www.bbc.com/ws/includes
+SIMORGH_INCLUDES_BASE_AMP_URL=https://news.files.bbci.co.uk
```
Comment (chris-hinds) [inline envConfig/live.env]: this is a duplicate of the `SIMORGH_PUBLIC_STATIC_ASSETS_ORIGIN` env var

File: envConfig/test.env [hunk 19]
```diff
@@ -4,6 +4,7 @@ SIMORGH_PUBLIC_STATIC_ASSETS_ORIGIN=https://news.test.files.bbci.co.uk
 SIMORGH_PUBLIC_STATIC_ASSETS_PATH=/include/articles/public/
 SIMORGH_ATI_BASE_URL=https://logws1363.ati-host.net?
 SIMORGH_TOGGLES_URL=https://toggles.test.api.bbci.co.uk
+SIMORGH_INCLUDES_BASE_AMP_URL=https://news.test.files.bbci.co.uk
```
Comment (chris-hinds) [inline envConfig/test.env]: same as above, we already specify this URL

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  // An amp-image query parameter on the include path indicates an AMP version of the include is available
+  const hasAmpImageQueryString = new Url(href, true).query['amp-image'];
+  return !!hasAmpImageQueryString;
+};
+
+const includeClassifier = ({ href, pathname }) => {
+  // This determines if the href has a leading '/'
+  const hrefTypePostion = () => (href.indexOf('/') === 0 ? 1 : 0);
+
+  // This checks if the supportedType is in the correct position of the href
```
Comment (chris-hinds) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/includeClassifier.js:22]: Nice to see lots of comments here 👍

**Review by HarveyPeachey — APPROVED** [review]
Great work on this! LGTM

**Comment by rhenshaw56** [comment]
@paruchurisilpa this won't need manual testing as it's just preprocessor logic, that would be handled in #6917

**Comment by paruchurisilpa** [comment]
@rhenshaw56 the PR that is resolved by this https://github.com/bbc/simorgh/issues/5989 needs manual testing, isn't that ready yet?

**Comment by rhenshaw56** [comment]
yeah, #5989 is the original issue and part of the work was split into #6917 (which would need manual testing)

**Comment by paruchurisilpa** [comment]
Thanks @rhenshaw56 will test once https://github.com/bbc/simorgh/issues/6917 is ready then..

**Comment by paruchurisilpa** [comment]
As mentioned above this can't be tested manually, will test this as part of https://github.com/bbc/simorgh/issues/6917, moving this to Ready to merge..

**Comment by amywalkerdev** [comment]
Bundle size for Russian is too large at 858 kB. Please update thresholds in './scripts/bundleSizeConfig.js'

**Comment by andrewscfc** [comment]
> As mentioned above this can't be tested manually, will test this as part of #6917, moving this to Ready to merge..

@paruchurisilpa @Simon2828 @rhenshaw56 
I think we could have performed an includes regression for this PR as it includes a significant refactor of the pre-processor rules used for all include types. Once we have the E2E/integration tests implemented: https://github.com/bbc/simorgh/issues/6827, I'd personally be comfortable making changes to includes and relying on these tests for regression coverage.
